### PR TITLE
Mosip 19185/machine key verification

### PR DESCRIPTION
**Pull request by abuabdul**

### CODE REVIEW

File: resident/resident-service/src/main/java/io/mosip/resident/constant/ApiName.java [hunk 1]
```diff
@@ -53,6 +53,10 @@
 
 	CREDENTIAL_CANCELREQ_URL,
 
+	TPMPUBLICKEY,
```
Comment (vishwa-vyom) [inline resident/resident-service/src/main/java/io/mosip/resident/constant/ApiName.java]: Rename this to PACKETSIGNPUBLICKEY

File: resident/resident-service/src/main/java/io/mosip/resident/constant/ResidentErrorCode.java [hunk 2]
```diff
@@ -40,7 +40,9 @@
 			"Exception while creating packet."),
 	INVALID_ID("RES-SER-29", "Invalid id"),
 	OTP_GENERATION_EXCEPTION("RES-SER-425", "while generating otp error is occured"),
-	POLICY_EXCEPTION("RES-SER-426", "while retrieving policy details error is occured");
+	POLICY_EXCEPTION("RES-SER-426", "while retrieving policy details error is occured"),
+	TPM_SIGNKEY_EXCEPTION("RES-SER-430", "Public sign key is not available from key manager"),
```
Comment (vishwa-vyom) [inline resident/resident-service/src/main/java/io/mosip/resident/constant/ResidentErrorCode.java]: rename to PACKET_SIGNKEY_EXCEPTION

File: resident/resident-service/src/main/java/io/mosip/resident/dto/MachineSearchRequestDTO.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+ * This class is used to search machine data from master data service.
+ *
+ * @author Abubacker Siddik
+ */
+
+@Data
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class MachineSearchRequestDTO implements Serializable {
```
Comment (vishwa-vyom) [inline resident/resident-service/src/main/java/io/mosip/resident/dto/MachineSearchRequestDTO.java]: like used in MachineSearchRequestDTO, can we use the BaseRestRequestDTO to be extended to this class and all other request and response classes ?

File: resident/resident-service/src/main/java/io/mosip/resident/dto/SignKeyRequestDTO.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ * This class is used to get TPM signkey from keymanager.
+ *
+ * @author Abubacker Siddik
+ */
+
+@Data
+@Builder
+public class SignKeyRequestDTO implements Serializable {
```
Comment (vishwa-vyom) [inline resident/resident-service/src/main/java/io/mosip/resident/dto/SignKeyRequestDTO.java]: Rename to PacketSignPublicKeyRequestDTO, follow the same naming for the relative things like response and internal classess

File: resident/resident-service/src/test/java/io/mosip/resident/test/service/ResidentServiceResUpdateTest.java [hunk 5]
```diff
@@ -128,7 +129,8 @@ public void setUp() throws BaseCheckedException, IOException {
 	}
 
 	@Test
-	public void reqUinUpdateSuccessTest() throws ResidentServiceCheckedException {
+	@Ignore
```
Comment (vishwa-vyom) [inline resident/resident-service/src/test/java/io/mosip/resident/test/service/ResidentServiceResUpdateTest.java:132]: Why are we ignoring this test case ?

File: resident/resident-service/src/main/java/io/mosip/resident/service/impl/ResidentServiceImpl.java [hunk 6]
```diff
@@ -588,6 +564,21 @@ public ResidentUpdateResponseDTO reqUinUpdate(ResidentUpdateRequestDto dto) thro
...
+            final String publicKey = getPublicKeyFromKeyManager();
+
+            // 2. Get all the machine with a prefix
+            MachineSearchResponseDTO machineSearchResponseDTO = searchMachineMasterWithMachineNamePrefix(residentMachinePrefix, publicKey);
+
+            // 3. Check the machine has same public key as that of tpm sign key
+            String machineId = getMachineId(machineSearchResponseDTO, publicKey);
+
+            // 4. Create new machine with a prefix in the master data
+            if (machineId == null) {
```
Comment (vishwa-vyom) [inline resident/resident-service/src/main/java/io/mosip/resident/service/impl/ResidentServiceImpl.java:571]: Please remove 1,2,3,4 in the above comments

File: resident/resident-service/src/main/java/io/mosip/resident/service/impl/ResidentServiceImpl.java [hunk 7]
```diff
@@ -719,4 +710,102 @@ private IdType getIdType(String individualType) {
...
 
+	public String getPublicKeyFromKeyManager() throws ApisResourceAccessException {
+		SignKeyRequestDTO signKeyRequestDto = SignKeyRequestDTO.builder().request(SignKeyRequestDTO.SignRequest.builder().serverProfile(SERVER_PROFILE_SIGN_KEY).build()).build();
+		SignKeyResponseDTO signKeyResponseDTO;
+		try {
+			HttpEntity<String> httpEntity = new HttpEntity<>(JsonUtils.javaObjectToJsonString(signKeyRequestDto));
+			signKeyResponseDTO = residentServiceRestClient.postApi(env.getProperty(ApiName.TPMPUBLICKEY.name()), MediaType.APPLICATION_JSON, httpEntity, SignKeyResponseDTO.class, tokenGenerator.getToken());
+			logger.debug(LoggerFileConstant.SESSIONID.toString(), LoggerFileConstant.USERID.toString(), SERVER_PROFILE_SIGN_KEY,
+					"ResidentServiceImpl::reqUinUpdate():: TPMPUBLICKEY POST service call ended with response data "
+							+ JsonUtils.javaObjectToJsonString(signKeyResponseDTO));
```
Comment (vishwa-vyom) [inline resident/resident-service/src/main/java/io/mosip/resident/service/impl/ResidentServiceImpl.java]: JsonUtils.javaObjectToJsonString is called twice, for performance reasons, put it to a variable in line 717 and reuse that.

File: resident/resident-service/src/main/java/io/mosip/resident/service/impl/ResidentServiceImpl.java [hunk 8]
```diff
@@ -719,4 +710,102 @@ private IdType getIdType(String individualType) {
...
+		} catch (Exception e) {
+			logger.error(LoggerFileConstant.SESSIONID.toString(), LoggerFileConstant.USERID.toString(), SERVER_PROFILE_SIGN_KEY,
+					"ResidentServiceImpl::reqUinUpdate():: TPMPUBLICKEY POST service call"
+							+ ExceptionUtils.getStackTrace(e));
+			throw new ApisResourceAccessException("Could not fetch public key from kernel keymanager", e);
+		}
+		return signKeyResponseDTO.getResponse().getPublicKey();
+	}
+
+	public MachineSearchResponseDTO searchMachineMasterWithMachineNamePrefix(String residentMachinePrefix, String publicKey) throws ApisResourceAccessException {
```
Comment (vishwa-vyom) [inline resident/resident-service/src/main/java/io/mosip/resident/service/impl/ResidentServiceImpl.java]: Rename this method as searchMachineInMasterService

File: resident/resident-service/src/main/java/io/mosip/resident/service/impl/ResidentServiceImpl.java [hunk 9]
```diff
@@ -719,4 +710,102 @@ private IdType getIdType(String individualType) {
...
+						.languageCode(utilities.getLanguageCode())
+						.build())
+				.build();
+		MachineSearchResponseDTO machineSearchResponseDTO;
+		try {
+			HttpEntity<String> httpEntity = new HttpEntity<>(JsonUtils.javaObjectToJsonString(machineSearchRequestDTO));
+			machineSearchResponseDTO = residentServiceRestClient.postApi(env.getProperty(ApiName.MACHINESEARCH.name()), MediaType.APPLICATION_JSON, httpEntity, MachineSearchResponseDTO.class, tokenGenerator.getToken());
+			logger.debug(LoggerFileConstant.SESSIONID.toString(), LoggerFileConstant.USERID.toString(), residentMachinePrefix,
+					"ResidentServiceImpl::reqUinUpdate():: MACHINESEARCH POST service call ended with response data "
+							+ JsonUtils.javaObjectToJsonString(machineSearchResponseDTO));
```
Comment (vishwa-vyom) [inline resident/resident-service/src/main/java/io/mosip/resident/service/impl/ResidentServiceImpl.java]: Avoid calling javaObjectToJsonString for performance reasons

File: resident/resident-service/src/main/java/io/mosip/resident/service/impl/ResidentServiceImpl.java [hunk 10]
```diff
@@ -719,4 +710,102 @@ private IdType getIdType(String individualType) {
...
+				List<MachineDto> machines = fetchedMachines.stream().filter(mac -> mac.getSignPublicKey().equals(publicKey)).collect(Collectors.toList());
+				if (!machines.isEmpty()) {
+					return machines.get(0).getId();
+				}
+			}
+		}
+		return null;
+	}
+
+	public String createNewMachineInMachineMaster(String residentMachinePrefix, String machineSpecId, String zoneCode, String regCenterId, String publicKey) throws ApisResourceAccessException {
```
Comment (vishwa-vyom) [inline resident/resident-service/src/main/java/io/mosip/resident/service/impl/ResidentServiceImpl.java]: Rename this to create createNewMachineInMasterService

**Review by vishwa-vyom — CHANGES_REQUESTED** [review]

File: resident/resident-service/src/test/java/io/mosip/resident/test/service/ResidentServiceResUpdateTest.java (hunk 5 above)
Comment (abuabdul) [reply resident/resident-service/src/test/java/io/mosip/resident/test/service/ResidentServiceResUpdateTest.java:132]: I tried to fix the test case when it is failing. residentserviceclient api has postApi with 5 arguments. They have mocked it. Each post API has different return type due to which i could not set up a proper mock. Hence the test is failing.

Comment (vishwa-vyom) [reply resident/resident-service/src/test/java/io/mosip/resident/test/service/ResidentServiceResUpdateTest.java:132]: Lets create this as a sub task in the code coverage to resolve this issue

**Review by vishwa-vyom — APPROVED** [review]

**Pull request merged**